### ISSUE TITLE
Completion not working with neovim/neovim-lspconfig

### ISSUE THREAD

**Issue by david0** [redhat-developer/yaml-language-server#370]
## Describe the bug
I'm trying to use yaml-language-server with neovim + [neovim-lspconfig](https://github.com/neovim/nvim-lspconfig#yamlls).

It looks like its not working at all.

## Expected Behavior
Completion for kubernetes yaml files etc

## Current Behavior
No completion is provided, e.g. if I try to complete a kubernetes yaml file:
![vim](https://user-images.githubusercontent.com/1685610/101676054-e944cd00-3a5a-11eb-8ed7-d1f3f64d9163.png)

there is an error in my lsp.log file:
```
[ ERROR ] 2020-12-09T19:55:06+0100 ] ...ovim/HEAD-8fb786e/share/nvim/runtime/lua/vim/lsp/rpc.lua:455 ]  "rpc"   "/Users/david/.cache/nvim/lspconfig/yamlls/node_modules/.bin/yaml-language-server"      "stderr"        "/Users/david/.cache/nvim/lspconfig/yamlls/node_modules/vscode-jsonrpc/lib/main.js:436
                        responsePromise.reject(new messages_1.ResponseError(error.code, error.message, error.data));
                                               ^

ResponseError: MethodNotFound\n    at handleResponse (/Users/david/.cache/nvim/lspconfig/yamlls/node_modules/vscode-jsonrpc/lib/main.js:436:48)
    at processMessageQueue (/Users/david/.cache/nvim/lspconfig/yamlls/node_modules/vscode-jsonrpc/lib/main.js:263:17)
    at Immediate.<anonymous> (/Users/david/.cache/nvim/lspconfig/yamlls/node_modules/vscode-jsonrpc/lib/main.js:247:13)
    at processImmediate (node:internal/timers:463:21) {
  code: -32601,
  data: undefined
}"
```
see my [full lsp.log](https://github.com/redhat-developer/yaml-language-server/files/5668075/lsp.log) for more detail

## Steps to Reproduce
 NVIM v0.5.0-8fb786e41 + recent 
1. install neovim (i have  NVIM v0.5.0-8fb786e41) 
2. ~/.config/nvim/init.vim:
```
call plug#begin('~/.local/share/nvim/plugged')
Plug 'neovim/nvim-lspconfig'
call plug#end()

autocmd FileType * set omnifunc=v:lua.vim.lsp.omnifunc completeopt=noinsert,menuone
lua << EOF
  local custom_lsp_attach = function(_, bufnr)
    vim.api.nvim_buf_set_option(bufnr, 'omnifunc', 'v:lua.vim.lsp.omnifunc')
  end
  require'lspconfig'.yamlls.setup {
    settings= {
      yaml = {
        schemas = {
          kubernetes='/test.yaml'
        }
      }
    },
    on_attach = custom_lsp_attach
  }
EOF
```
3. open a kubernetes yaml file and try omnicompletion via CTRL-X CTRL-O 
## Environment
- [ ] Windows
- [x] Mac
- [ ] Linux
- [ ] other (please specify)

**Comment by JPinkney** [comment]
Does hover work by any chance? If my hunch is correct then I don't think it would be. It seems like the `workspaceFolders` is trying to dynamically register itself when even though the client says `dynamicRegistration` is not supported. I believe this error would make none of the features work (since it can't pickup the workspaceFolder). Most likely it's an issue with: https://github.com/redhat-developer/yaml-language-server/blob/master/src/server.ts#L419 not checking the status of dynamicRegistration

**Comment by david0** [comment]
Hover does not work either:
```
E5108: Error executing lua ...HEAD-8fb786e/share/nvim/runtime/lua/vim/lsp/handlers.lua:302: RPC[Error] code_name = MethodNotFound, message = "method textDocument/hover is not supported by any of the servers registered for the current buffer"   
```

Yes it really looks to me like nothing is working at all.

**Comment by david0** [comment]
Yeah setting hasWorkspaceFolderCapability to false made it work! 🎉 

I guess a generic solution would be:
```diff
diff --git a/src/server.ts b/src/server.ts
index d87390f..ccfc77a 100755
--- a/src/server.ts
+++ b/src/server.ts
@@ -438,7 +438,7 @@ connection.onInitialize(
 );
 
 connection.onInitialized(() => {
-  if (hasWorkspaceFolderCapability) {
+  if (hasWorkspaceFolderCapability && clientDynamicRegisterSupport) {
     connection.workspace.onDidChangeWorkspaceFolders((changedFolders) => {
       workspaceFolders = workspaceFoldersChanged(workspaceFolders, changedFolders);
     });
```

**Comment by JPinkney** [comment]
I think we might have to do something like change: https://github.com/redhat-developer/yaml-language-server/blob/master/src/server.ts#L419 to be:
```hasWorkspaceFolderCapability = !!(capabilities.workspace && capabilities.workspace.workspaceFolders);```
instead.

Technically `clientDynamicRegisterSupport` is badly named because it actually only checks if dynamicRegistration for the formatter is allowed. If in the off chance that a client supports dynamic registration for formatter but doesn't support dynamic registration for workspaceFolders I think your code might fail.

If my small changes work feel free to open a PR (or I can its upto you) and I can get those changes merged! Once the CI runs when it's merged they changes should be available soon after!

**Comment by david0** [comment]
Unfortunately your change does not work for me. Neovim claims it supports these features ([full list](https://github.com/redhat-developer/yaml-language-server/files/5668833/capabilities.txt))

I believe it tracks down to https://github.com/neovim/neovim/issues/13448.
Depending how long the resolution there takes, maybe misusing some (or all) of the other dynamicRegistration properties would be a workaround?

**Comment by david0** [comment]
This can be closed because of the improvements implemented with neovim/nvim-lspconfig#471 and #378
